File: examples/dcps/Ownership/java/src/OwnershipDataPublisher.java
/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to TO_YEAR PrismTech
 *   Limited, its affiliated companies and licensors. All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
/************************************************************************
 *
 * Copyright (c) 2007
 * PrismTech Ltd.
 * All rights Reserved.
 *
 * LOGICAL_NAME:    OwnershipDataPublisher.java
 * FUNCTION:        OpenSplice example code.
 * MODULE:          Tutorial for the Java programming language.
 * DATE             April 2009.
 ************************************************************************
 *
 * This file contains the implementation for the 'OwnershipDataPublisher' executable.
 *
 ***/

import DDS.*;
import OwnershipData.*;

public class OwnershipDataPublisher {
	DDSEntityManager mgr = new DDSEntityManager();
	DataWriter dWriter;
	StockDataWriter OwnershipDataDataWriter;
	long userHandle;
	Stock m_instance;
	int status;

	void initPublisher(String pub, int strength) {
		mgr = new DDSEntityManager();
		String partition_name = "Ownership example";
		mgr.createParticipant(partition_name);
		// create type
		StockTypeSupport st = new StockTypeSupport();
		mgr.registerType(st);
		// create Topic
		String topic_name = "OwnershipStockTracker";
		mgr.createTopic(topic_name);
		// create Publisher
		mgr.createPublisher();

		// create DataWriter
		mgr.createWriter(strength);

		dWriter = mgr.getWriter();
		OwnershipDataDataWriter = StockDataWriterHelper.narrow(dWriter);

		m_instance = new Stock();

		m_instance.ticker = "MSFT";
		m_instance.price = 0;
		m_instance.publisher = pub;
		m_instance.strength = strength;
		userHandle = OwnershipDataDataWriter.register_instance(m_instance);

	}

	OwnershipDataPublisher(String pub, int strength) {
		initPublisher(pub, strength);
	}

	void publishEvent(float price, String pub) {
		m_instance.price = price;
		m_instance.publisher = pub;
		OwnershipDataDataWriter.write(m_instance, userHandle);
	}

	void dispose() {

		mgr.getPublisher().delete_datawriter(OwnershipDataDataWriter);

		/* Remove the Publisher. */
		mgr.deletePublisher();

		/* Remove the Topics. */
		mgr.deleteTopic();

		/* Remove Participant. */
		mgr.deleteParticipant();

	}

	public static void main(String args[]) {

		System.out.println(args.length);
		if (args.length < 4) {
			System.err
					.println("*** [Publisher] usage : Publisher <publisher_name> <ownership_strength> <nb_iterations> <stop_subscriber_flag>");
			System.exit(-1);
		}

		OwnershipDataPublisher pub;
		String publisher_name = args[0];
		int ownership_strength = Integer.parseInt(args[1]);
		int nb_iteration = Integer.parseInt(args[2]);
		pub = new OwnershipDataPublisher(publisher_name, ownership_strength);
		boolean stop_subscriber = (Integer.parseInt(args[3]) == 1);

		// Publisher publishes the prices in dollars
		System.out.println("=== [Publisher] Publisher " + publisher_name
				+ " with strength : " + ownership_strength);
		System.out.println(" / sending " + nb_iteration + " prices ..."
				+ " stop_subscriber flag=" + args[3]);
		// The subscriber should display the prices sent by the publisher with
		// the highest ownership strength
		float price = 10.0f;
		for (int x = 0; x < nb_iteration; x++) {
			pub.publishEvent(price, publisher_name);
			// Sleep(delay_200ms);
			try {
				Thread.currentThread().sleep(200);// sleep for 200 ms
			} catch (InterruptedException ie) {
			}
			price = price + 0.5f;
		}
		// Sleep(delay_2s);
		try {
			Thread.currentThread().sleep(2000);// sleep for 2000 ms
		} catch (InterruptedException ie) {
		}
		if (stop_subscriber) {
			// send a price = -1 to stop subscriber
			price = -1.0f;
			System.out.println("=== Stopping the subscriber");
			pub.publishEvent(price, publisher_name);
		}
		pub.dispose();
	}
}

File: examples/dcps/Ownership/c/src/OwnershipDataPublisher.c
/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to TO_YEAR PrismTech
 *   Limited, its affiliated companies and licensors. All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

/************************************************************************
 * LOGICAL_NAME:    OwnershipDataPublisher.c
 * FUNCTION:        Publisher's main for the OwnershipData OpenSplice example.
 * MODULE:          OpenSplice OwnershipData example for the C programming language.
 * DATE             September 2010.
 ************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#ifndef _WIN32
#include <unistd.h>
#else
#include "os_stdlib.h"
#endif
#include <string.h>
#include "dds_dcps.h"
#include "CheckStatus.h"
#include "DDSEntitiesManager.h"

#include "example_main.h"

void usage()
{
   fprintf(stderr, "\n\n *** ERROR");
   fprintf(stderr, "\n *** [Publisher] usage: \n \t Publisher <publisher_name> <ownership_strength> <nb_iterations>  <is_stopping_subscriber = 0 OR 1>");
   exit(-1);
}

int OSPL_MAIN (int argc, char *argv[])
{
   char* publisher_name;
   int i, sampleIndex, ownership_strength, nb_iteration;
   c_bool isStoppingSubscriber;
   float price;
   os_time os_delay200 = { 0, 200000000 };
   os_time os_delay2000 = { 2, 0 };


   DDS_Publisher OwnershipDataPublisher;
   DDS_DataWriter OwnershipDataDataWriter;
   OwnershipData_Stock* OwnershipDataSample;
   DDS_InstanceHandle_t userHandle = NULL;
   const char ticker [] = "MSFT";
   DDS_unsigned_long stringLength = sizeof(ticker) - 1;

   // usage : Publisher.exe <publisher_name> <ownership_strength> <nb_iterations>
   if( argc < 5 )
   {
      usage();
   }
   printf("\n\n Starting OwnershipDataPublisher...");
   printf("\n\n Parameters are:");
   for( i = 1; i < argc ; ++i )
   {
      printf("\n %d: %s", i, argv[i]);
   }

   publisher_name = argv[1];
   ownership_strength = atoi(argv[2]);
   nb_iteration = atoi(argv[3]);
   isStoppingSubscriber =  argv[4][0] == '1' ? TRUE : FALSE;

   createParticipant("Ownership example");

   // Register Stock Topic's type in the DDS Domain.
   g_StockTypeSupport = (DDS_TypeSupport) OwnershipData_StockTypeSupport__alloc();
   checkHandle(g_StockTypeSupport, "OwnershipData_StockTypeSupport__alloc");
   registerStockType(g_StockTypeSupport);
   // Create Stock Topic in the DDS Domain.
   g_StockTypeName = OwnershipData_StockTypeSupport_get_type_name(g_StockTypeSupport);
   g_StockTopic = createTopic("OwnershipStockTracker", g_StockTypeName);
   DDS_free(g_StockTypeName);
   DDS_free(g_StockTypeSupport);

   // Create the Publisher's in the DDS Domain.
   OwnershipDataPublisher = createPublisher();

   // Request a Writer from the the Publisher.
   OwnershipDataDataWriter = createDataWriter(OwnershipDataPublisher, g_StockTopic, FALSE, ownership_strength);

   // Publish a Stock Sample reflecting the state of the Msg DataWriter.
   OwnershipDataSample = OwnershipData_Stock__alloc();

   OwnershipDataSample->ticker = DDS_string_alloc(stringLength);
   snprintf(OwnershipDataSample->ticker, stringLength + 1, "%s", ticker);
   OwnershipDataSample->price = 0.0;
   OwnershipDataSample->publisher = DDS_string_dup(publisher_name);
   OwnershipDataSample->strength = ownership_strength;

   //Publisher publishes the prices in dollars
   printf("\n=== [Publisher] Publisher %d with strength : ", ownership_strength);
   printf("\n / sending %d prices...", nb_iteration);
   // The subscriber should display the prices sent by the publisher with the highest ownership strength
   price = 10.0f;
   for( sampleIndex = 0; sampleIndex < nb_iteration ; ++sampleIndex )
   {
      OwnershipDataSample->price = price;
      writeStockSample(OwnershipDataDataWriter, userHandle, OwnershipDataSample);
      os_nanoSleep(os_delay200);
      price = price + 0.5f;
   }
   if( isStoppingSubscriber == TRUE )
   {
      // This special price gives the end signal to the Subscriber:
      OwnershipDataSample->price =  (DDS_float) -1.0f;
      writeStockSample(OwnershipDataDataWriter, userHandle, OwnershipDataSample);
   }
   /* A short sleep ensures time is allowed for the sample to be written to the network.
   If the example is running in *Single Process Mode* exiting immediately might
   otherwise shutdown the domain services before this could occur */
   os_nanoSleep(os_delay2000);

   // Cleanup DDS from the created Entities.
   deleteDataWriter(OwnershipDataPublisher, OwnershipDataDataWriter);
   deletePublisher(OwnershipDataPublisher);
   deleteTopic(g_StockTopic);
   deleteParticipant();

   // Cleanup C allocations,
   // recursively freeing the allocated structures and sequences using the OpenSplice API.
   DDS_free(OwnershipDataSample);

   // Print out an empty line, just to let behind a clean new line for the shell..
   printf("\n\r");
   return 0;
}

